src/device.rs

```diff
@@ -646,36 +648,44 @@ pub extern "C" fn wgpuDeviceCreateSwapChain(
     device: id::DeviceId,
     surface: id::SurfaceId,
     descriptor: &native::WGPUSwapChainDescriptor,
-) -> id::SwapChainId {
-    let desc = wgt::SwapChainDescriptor {
+) -> id::SurfaceId {
+    // The swap chain API of wgpu-core (and WebGPU) has been merged into the surface API,
+    // so this gets a bit weird until the webgpu.h changes accordingly.
+    let config = wgt::SurfaceConfiguration {
         usage: wgt::TextureUsages::from_bits(descriptor.usage).unwrap(),
         format: conv::map_texture_format(descriptor.format).expect("Texture format not defined"),
         width: descriptor.width,
         height: descriptor.height,
         present_mode: conv::map_present_mode(descriptor.presentMode),
     };
-    let (id, error) =
-        gfx_select!(device => GLOBAL.device_create_swap_chain(device, surface, &desc));
+    let error = gfx_select!(device => GLOBAL.surface_configure(surface, device, &config));
     if let Some(error) = error {
-        panic!("Failed to create swapchain: {}", error);
+        panic!("Failed to configure surface: {}", error);
     }
-
-    id
+    surface // swap chain_id == surface_id
 }
 
 #[no_mangle]
 pub extern "C" fn wgpuSwapChainGetCurrentTextureView(
-    swap_chain: id::SwapChainId,
+    swap_chain: id::SurfaceId,
 ) -> Option<id::TextureViewId> {
-    gfx_select!(swap_chain => GLOBAL.swap_chain_get_current_texture_view(swap_chain, PhantomData))
+    let surface_id = swap_chain;
+    let device_id = surface_id; // TODO get device from surface?
+    let result =
+        gfx_select!(device_id => GLOBAL.surface_get_current_texture(surface_id, PhantomData));
+    let texture = result
         .expect("Unable to get swap chain texture view")
-        .view_id
+        .texture_id
+        .unwrap();
+    let desc = wgc::resource::TextureViewDescriptor::default();
+    Some(gfx_select!(texture => GLOBAL.texture_create_view(texture, &desc, PhantomData)).0)
 }
 
 #[no_mangle]
-pub extern "C" fn wgpuSwapChainPresent(swap_chain: id::SwapChainId) {
-    //TODO: Header does not return swap chain status?
-    gfx_select!(swap_chain => GLOBAL.swap_chain_present(swap_chain))
+pub extern "C" fn wgpuSwapChainPresent(swap_chain: id::SurfaceId) {
+    let surface_id = swap_chain;
+    let device_id = surface_id; // TODO get device from surface?
+    gfx_select!(device_id => GLOBAL.surface_present(surface_id))
         .expect("Unable to present swap chain");
 }
```

Comment:
I'm stuck here. From what I understand, the `gfx_select!` macro wants an argument (before the arrow operator) that is an id that can be associated with the backend in use. The `surface_id` does not work; it panics with "Unexpected backend Empty". I've seen [this code in wgpu-rs](https://github.com/gfx-rs/wgpu/blob/9bc59084926dc66a29bfffdc0a960d7cebbf2e62/wgpu/src/backend/direct.rs#L917-L920) to obtain the device from the surface, but I don't know how to reproduce that here. I'm not even sure that's the way to go TBH :)

@kvark Could you please provide some pointers to help me move forward?

Reply:
That's indeed a bit of a problem here, and I'm not too happy about how wgpu-rs solves this. It's able to get away because it can have arbitrary payloads to the objects like `Surface`. Can wgpu-native have a payload to `SurfaceId`? If it can't, we need to figure out how to dispatch this internally, i.e. how to implement `fn surface_present` on wgpu-core side that wouldn't need the generic parameter.

Reply:
I seem to have made it work using a global HashMap and `lazy_static`. My downstream tests and examples work!